hacks/prune-rhcos-snapshots/prune-rhcos-snapshots.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python3
+import os
+import re
+
+import boto3
+import openshift as oc
+import re
+import json
+import yaml
+from typing import Dict, List, Set
+import datetime
+import urllib3
+import pathlib
+import subprocess
+
+# Set to "true" if an AMI is selected for garbage collection
+AMI_TAG_KEY_GARBAGE_COLLECT = 'garbage_collect'
+
+# Set to "true" if an AMI is mentioned in installer reference
+AMI_TAG_KEY_PRODUCTION = 'production'
+
+
+def create_recycle_bin_rule(client, res_type: str):
+    response = client.create_rule(
+        RetentionPeriod={
+            'RetentionPeriodValue': 60,
+            'RetentionPeriodUnit': 'DAYS'
+        },
+        Tags=[
+            {
+                'Key': 'Name',
+                'Value': f'rhcos-preserve-{res_type}'
+            },
+        ],
+        Description=f'Help protect RHCOS {res_type} from accidental deletion',
+        ResourceType=res_type,
+    )
+    print(f'Created new retention rule: {response["Identifier"]}')
+
+
+Commit = str
+Region = str
+AmiId = str
+
+if __name__ == '__main__':
+
+    gopath = pathlib.Path(os.getenv('GOPATH'))
+    if not gopath.is_dir():
+        print('Set GOPATH env before running')
+        exit(1)
+
+    installer_git_path = gopath.joinpath('src/github.com/openshift/installer')
+    if not installer_git_path.is_dir():
+        print(f'Requires recent clone of github.com/openshift/installer in GOPATH: {str(installer_git_path)}')
+        exit(1)
+
+    http = urllib3.PoolManager()
+
+    amis_in_use: Set = set()
+
+    git_depth = 0
+    while True:
+        installer_commit: Commit = f'upstream/master~{git_depth}'
+        print(f'  Processing installer commit: {installer_commit}')
+        try:
+            gitshow_process = subprocess.run(['git', '-C', str(installer_git_path), 'show', f'{installer_commit}:data/data/coreos/rhcos.json'], capture_output=True)
+            gitshow_process.check_returncode()
+        except:
+            gitshow_process = subprocess.run(['git', '-C', str(installer_git_path), 'show', f'{installer_commit}:data/data/rhcos.json'], capture_output=True)
+            if gitshow_process.returncode != 0:
+                print(f'Failed to find rhcos.json at depth {git_depth}')
+                break
+
+        # Otherwise gitshow process contains a file like https://github.com/openshift/installer/blob/01adff5d629bb403a6cf49f9d44070762ef77e93/data/data/coreos/rhcos.json
+        # or older format like https://github.com/openshift/installer/blob/bfdba2d19e3fa1105bab93aa4cce8bb7f44b4b2c/data/data/rhcos.json
+        amis_in_file = list(re.findall('ami-[a-z0-9]+', str(gitshow_process.stdout)))
+        amis_in_use.update(amis_in_file)
+        git_depth += 1
```

Comment:
Instead of visiting every commit on the branch, it'd be more efficient to ask `git` for only the commits which touch the files we care about and then only visit those. Something like:

```
git log --format=%H <branch> -- data/data/coreos/rhcos.json data/data/rhcos.json
```

Reply:
I like it. Updated.